### PR TITLE
Update: warn when used with not supported Node versions

### LINKED ISSUE #4501: Make the node requirements clear
**Do you want to request a *feature* or report a *bug*?**

feature

**What is the current behavior?**

yarn will fail on unsupported versions of node

**If the current behavior is a bug, please provide the steps to reproduce.**

**What is the expected behavior?**

spit out some warning before failing

**Please mention your node.js, yarn and operating system version.**

< ^4.8.0 and <= 5.7.0

see also https://github.com/yarnpkg/website/issues/646 for more information

### PR DESCRIPTION
**Pull request by BYK**
**Summary**

Fixes #4501, refs #4490, refs #4284. Yarn now warns when it
detects it is running in a Node version that is not fully
supported and warns the user about this. This is different than
the hard Node 4+ check in the entry file since in that case,
Yarn wouldn't run at all due to syntax incompatibilities. This
warning is to signal that users may encounter unexpected errors
but are allowed to use Yarn if they wish. It also adds a new
flag to suppress this warning: `--no-node-version-check`.

**Test plan**

Since we cannot add unsupported Node versions to our CI and
spoof the Node version internally, this has to be tested
manually, which I did.

### CODE REVIEW

**Comment by SimenB** [comment]
You can mock the node version if you really want to.

https://github.com/SimenB/node-version-check/blob/d761da1b4fa2bac2845b4ba31c525817f4d31755/test/lt8.test.js#L4-L26

**Comment by BYK** [comment]
@SimenB yes but that requires us to run Yarn from inside the same process that mocks the version which is not easy or may not be possible with the CLI entry point.

File: src/cli/index.js [hunk 1]
```diff
@@ -98,6 +100,7 @@ export function main({
     'prepend the node executable dir to the PATH in scripts',
     boolify,
   );
+  commander.option('--no-node-version-check', 'do not warn when using a potentially unsupported Node version');
```
Comment (arcanis) [inline src/cli/index.js]: That's a long command name :D

Comment (Haroenv) [reply src/cli/index.js]: `--no-node-version-check-or-else-youll-be-fired`

**Comment by mkj28** [comment]
Trying to understand why was node 7 dropped from support? Cannot find the reason in the documents linked here.

**Comment by BYK** [comment]
@mkj28 I think I dropped it unintentionally since it is not listed anywhere and it is not LTS. We do not test anything with Node 7 so we do not commit to supporting it. This doesn't mean it is going to break.

**Pull request merged**